### ISSUE TITLE
Siamese Vs. Single network in training

### ISSUE THREAD

**Issue by wenjie710** [marialeyvallina/generalized_contrastive_loss#4]
Hi, I have read your paper and the code. The work is cool and fantastic. 
  However, I am confused about the siamese network here. It is said in the paper that these two networks share weights and the same structure, so what is the difference if the tensors are input only once using single network? Like, how about we concatenate these two tensors into one batch and input it to the network? 
  Could you please provide some ablation study on Siamese Vs. Single  network in MSLS? Thanks a lot.

**Comment by marialeyvallina** [comment]
Hi @wenjie710 
We follow a traditional siamese network architecture, which is essentially one network at which we feed two images (x_i and x_j), with a certain similarity label (that we denote psi). So the pseudo-code pipeline is as follows:

We first forward two images. Please note that although the network and its weights are the same, the inputs are different, and the outputs are therefore different too.
```
out_i=network.forward(x_i)
out_j=network.forward(x_j)
```
We calculate the loss:
```
error = loss.forward(out_i, out_j, psi)
```

And finally, we backpropagate and update the network weights.
```
error.backward()
network.update()

```
This is also illustrated in [our paper,](https://arxiv.org/pdf/2103.06638.pdf) Fig. 2.

The batches are constructed with pairs of images, so for a batch of size N you'll have batch_xi of shape (N, 3, w, h) and batch_xj of shape (N,3,w,h), which you will forward to the network following the pipeline above.

To the best of my knowledge, all deep learning work in visual place recognition or image retrieval for MSLS or other datasets follows a siamese or triplet architecture, so I am not aware of the existence of any ablation study of Siamese Vs. Single network. 

Please do not hesitate to send us an email if you have further questions.

**Comment by wenjie710** [comment]
"The batches are constructed with pairs of images, so for a batch of size N you'll have batch_xi of shape (N, 3, w, h) and batch_xj of shape (N,3,w,h), which you will forward to the network following the pipeline above."

OK. Got it. So I guess there is no difference if we input (2N, 3, w, h) and seperate the output (2N, 3, ww, hh) into (N, 3, ww, hh) and (N, 3, ww, hh).

**Comment by mlopezantequera** [comment]
I agree there's no difference
